components/tenant-mgt/org.wso2.carbon.tenant.mgt/src/main/java/org/wso2/carbon/tenant/mgt/services/TenantMgtImpl.java

```diff
@@ -199,6 +199,98 @@ public void deactivateTenant(String tenantUniqueID) throws TenantMgtException {
+            } catch (UserStoreException e) {
+                throw new TenantManagementServerException("Error while getting the tenant with the tenant unique id :" +
+                        tenantUniqueIdentifier + " .", e);
+            }
+
+            if (tenant != null) {
+                int tenantId = tenant.getId();
+                String tenantDomain = tenant.getDomain();
+
+                try {
+                    if (log.isDebugEnabled()) {
+                        log.debug(String.format(
+                                "Starting tenant deletion for domain: %s and tenant id: %d from the system",
+                                tenantDomain, tenantId));
+                    }
+
+                    ServerConfigurationService
+                            serverConfigurationService = TenantMgtServiceComponent.getServerConfigurationService();
+
+                    if (Boolean.parseBoolean(serverConfigurationService.getFirstProperty("Tenant.TenantDelete"))) {
+                    /*
+                     * TODO: 2/7/19 We need to fix listeners to enable this by default
+                     * Refer - https://github.com/wso2-support/carbon-multitenancy/issues/35
+                     */
+                        if (Boolean.parseBoolean(serverConfigurationService
+                                .getFirstProperty("Tenant.ListenerInvocationPolicy.InvokeOnDelete"))) {
+                            notifyTenantDeletion(tenantId);
+                        } else {
+                            if (log.isDebugEnabled()) {
+                                log.debug(
+                                        "Tenant.ListenerInvocationPolicy.InvokeOnDelete flag is not set to true in carbon.xml. Listener invocation ignored.");
+                            }
+                        }
+
+                        TenantMgtUtil.deleteWorkernodesTenant(tenantId);
+
+                        if (TenantMgtServiceComponent.getBillingService() != null) {
+                            TenantMgtServiceComponent.getBillingService().deleteBillingData(tenantId);
+                        }
+
+                        TenantMgtUtil.unloadTenantConfigurations(tenantDomain, tenantId);
+                        TenantMgtUtil.deleteTenantRegistryData(tenantId);
+                        TenantMgtUtil.deleteTenantDir(tenantId);
+                        tenantManager.deleteTenant(tenantUniqueIdentifier);
+                        log.info(String.format("Deleted tenant with domain: %s and tenant id: %d from the system.",
+                                tenantDomain, tenantId));
+                    } else {
+                        if (log.isDebugEnabled()) {
+                            log.debug(
+                                    "Tenant.TenantDelete flag is not set to true in carbon.xml. Tenant will not be deleted.");
+                        }
+                    }
+                } catch (Exception e) {
```

Comment:
Do we need to catch the Exception type here? Aren't there specific expectations types?